### PR TITLE
Add initial whitelist for Stargate Query

### PR DESCRIPTION
**Pull request by mattverse**
Closes: https://github.com/osmosis-labs/osmosis/issues/2433

## What is the purpose of the change

This PR adds initial list of queries to whitelist for Stargate query. 

Queries using iterators to iterate through kv store is safe, while a query returning a non sorted array, or a query using randomness is considered unsafe. 

## Brief Changelog

- Add initial whitelist for stargate queries

### CODE REVIEW

File: wasmbinding/stargate_whitelist.go [hunk 1]
```diff
@@ -17,8 +29,83 @@ import (
 var StargateWhitelist sync.Map
 
 func init() {
+	// cosmos-sdk queries
+
+	// auth
 	StargateWhitelist.Store("/cosmos.auth.v1beta1.Query/Account", &authtypes.QueryAccountResponse{})
```
Comment (alexanderbez) [inline wasmbinding/stargate_whitelist.go:35]: I wish there was a way to infer these w/o hard-coding them :-/

**Review by alexanderbez — APPROVED** [review]

File: wasmbinding/stargate_whitelist.go [hunk 2]
```diff
@@ -17,8 +29,83 @@ import (
...
+	StargateWhitelist.Store("/cosmos.bank.v1beta1.Query/Params", &banktypes.QueryParamsResponse{})
+	StargateWhitelist.Store("/cosmos.bank.v1beta1.Query/SupplyOf", &banktypes.QuerySupplyOfResponse{})
+
+	// distribution
+	StargateWhitelist.Store("/cosmos.distribution.v1beta1.Query/Params", &distributiontypes.QueryParamsResponse{})
+	StargateWhitelist.Store("/cosmos.distribution.v1beta1.Query/DelegatorWithdrawAddress", &distributiontypes.QueryDelegatorWithdrawAddressResponse{})
+	StargateWhitelist.Store("/cosmos.distribution.v1beta1.Query/ValidatorCommission", &distributiontypes.QueryValidatorCommissionResponse{})
+
+	// evidence
+	StargateWhitelist.Store("/cosmos.evidence.v1beta1.Query/Evidence", &evidencetypes.QueryEvidenceResponse{})
```
Comment (ValarDragon) [inline wasmbinding/stargate_whitelist.go:50]: Lets remove this, I don't want any backwards compatability guarantees on this one lol

**Review by ValarDragon — APPROVED** [review]

File: wasmbinding/stargate_whitelist.go (hunk 2 above)
Comment (ValarDragon) [inline wasmbinding/stargate_whitelist.go:50]:
```suggestion
```

File: wasmbinding/stargate_whitelist.go [hunk 3]
```diff
@@ -5,8 +5,20 @@ import (
 
 	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
 	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	distributiontypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+	evidencetypes "github.com/cosmos/cosmos-sdk/x/evidence/types"
```
Comment (ValarDragon) [inline wasmbinding/stargate_whitelist.go:9]:
```suggestion
```

**Comment by p0mvn** [comment]
Why do we not have `Query/AllBalances` here anymore? It was present in https://github.com/osmosis-labs/osmosis/pull/2353

**Comment by mattverse** [comment]
@p0mvn They were added temporarily as an example, removed it for now with safety concerns, and we need more in depth review on the query!

File: wasmbinding/stargate_whitelist.go (hunk 1 above)
Comment (mattverse) [reply wasmbinding/stargate_whitelist.go:34]: https://github.com/osmosis-labs/osmosis/issues/2627

**Comment by ValarDragon** [comment]
Also all balances generally shouldn't be used, as it can have unbounded gas. Its a common footgun to not realize this

File: wasmbinding/stargate_whitelist.go (hunk 1 above)
Comment (marbar3778) [reply wasmbinding/stargate_whitelist.go:34]: This should be possible with the chnages being done in the sdk to support deterministic queries.

Comment (ThanhNhann) [reply wasmbinding/stargate_whitelist.go:34]: hi @marbar3778, can you give me the link about the changes you said, thanks!

Comment (marbar3778) [reply wasmbinding/stargate_whitelist.go:34]: https://github.com/cosmos/cosmos-sdk/issues/13041

**Pull request merged**